Add rpc.async_execution support for rpc.remote on script functions

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39758 Add rpc.async_execution support for rpc.remote on script functions**

Differential Revision: [D21963789](https://our.internmc.facebook.com/intern/diff/D21963789)

mrshenli committed:
Add rpc.async_execution support for rpc.remote on script functions

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/torchscript_functions.cpp:
```diff
@@ -17,9 +17,9 @@ c10::intrusive_ptr<c10::ivalue::Future> rpcTorchscript(
+    const bool isAsyncFunction) {
   auto scriptCall = std::make_unique<ScriptCall>(
-      qualifiedName, std::move(stack), asyncFunction);
+      qualifiedName, std::move(stack), isAsyncFunction);
```
`isAsyncFunction` should be named to `isAsyncExecution`, will update in the next version.

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -344,29 +344,55 @@ void RequestCallbackImpl::processRpc(
+                  }
+                  postProcessing();
+                });
+          } catch (const std::exception& e) {
```
Curious, what errors can happen here? Is it when the user async function does not return a future or is not properly decorated?

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -344,29 +344,55 @@ void RequestCallbackImpl::processRpc(
+      } else {
+        scriptPostProcessing = [ownerRRef, postProcessing](
+            const c10::intrusive_ptr<c10::ivalue::Future>& jitFuture) {
+          try {
```
Maybe consider deduping this block (from try to postProcessing) with the one at the top? Though it's only a small block of code, so maybe not too important.

rohan-varma commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -1105,3 +1105,46 @@ def test_async_function_wrong_decorator_order(self):
+            rrefs.append(
+                rpc.remote(
+                    dst1,
+                    async_add,
```
Unless I'm missing something, `async_add` should be annotated with torch.jit.script to test with a script function right?

rohan-varma commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -1105,3 +1105,46 @@ def test_async_function_wrong_decorator_order(self):
+                worker_name((self.rank + 1) % self.world_size),
+                async_wrong_type
+            )
+            rref.to_here()
```
Maybe take the `rpc.remote()` outside of the assert block, because we should be sure that to_here() is what raises right?

mrshenli committed:
Update on "Add rpc.async_execution support for rpc.remote on script functions"


Differential Revision: [D21963789](https://our.internmc.facebook.com/intern/diff/D21963789)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_execution support for rpc.remote on script functions"


Differential Revision: [D21963789](https://our.internmc.facebook.com/intern/diff/D21963789)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_execution support for rpc.remote on script functions"


Differential Revision: [D21963789](https://our.internmc.facebook.com/intern/diff/D21963789)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_execution support for rpc.remote on script functions"


Differential Revision: [D21963789](https://our.internmc.facebook.com/intern/diff/D21963789)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Thanks for addressing the previous comments I had!
